Implement monitoring-related functions

Aretno opened the pull request:
Hi everybody, this PR is to discuss and add the monitoring software to IC. As I have nearly zero idea in which would be the best way to add this to our scheme and before doing too much work that could not be useful I'd rather ask. 

First, I'd like to talk about the writer, starting from @andLaing work at io/hist_io.py I modified it so do not have to declare a writer instance for every histogram. In the meanway I added the information of under/over flow. It would be the following function (of course, it should be made clean, optimized with the help of those who know the ways of the python, etc., this is just a rough and starting base to start the discussion):

```
def hist_writer_var(file, *, compression = 'ZLIB4'):
    
    def write_hist(group_name : 'options: HIST, HIST2D',
                   table_name : 'histogram name',
                   entries    : 'np.array: RWF, CWF, SiPM',
                   bin_centres: 'np.array of bin centres',
                   out_of_range: 'np.array lenght=2 with events out of range'):
        
        try:                       hist_group = getattr          (file.root, group_name)
        except tb.NoSuchNodeError: hist_group = file.create_group(file.root, group_name)

        try:
            hist_table = getattr(hist_group, table_name)
        except tb.NoSuchNodeError:
            file.create_array(hist_group, table_name, entries)
            file.create_array(hist_group, table_name+'_bins', bin_centres)
            file.create_array(hist_group, table_name+'_outRange', out_of_range)
                    
    return write_hist
```

It basically creates an array with the variable name containing the entries, another with the name+'_bins' to containt the binning and another named with the variable name + '_outRange' to containt the under/over flow. Are you ok with something like this or is a terrible idea?

Moving on, my idea was to get the variables we want to plot on an file by file basis, plot them and refresh the variables so we do not end up with reaaaally long arrays in case we want to run the monitoring through a bunch of files. I know it will probably be run always on individual files but I though it would be better if we could accommodate the code for other cases.

For this I created a custom class MonPar (Monitoring Parameters, temporal name) which is initialized with an array of parameter names that it will contain. I also create an instance of that class with the binning of each parameter and another for the binned values, the output of the histogram. To bin the data I use a modification of the bin_waveform function that Andrew did (in one of his branches), so it returns the data binned in the bins given and the under/overflows, it would be the following:

```
def bin_data(data, histbins): #Copiado de Andrew en invisible_cities/calib/sipmpdf.py
    """
    Bins the current event data and adds it
    to the file level bin array
    """
    bin_data = lambda x: [np.histogram(x, histbins)[0], np.array([len(x[x<histbins[0]]), len(x[x>histbins[-1]])])]
    return np.apply_along_axis(bin_data, 0, data)
```
And here you have the MonPar class I was talking about:

```
class MonPar:
    def __init__(self, attrNames, values=None):
        for attribute in attrNames:
            if values:
                setattr(self, attribute, values[attribute])
            else:
                setattr(self, attribute, []) 
        self.attrNames = attrNames

    def BinValues(self, bins):
        for attribute in self.attrNames:
            attrVal = getattr(self, attribute)
            attrBin = getattr(bins, attribute)
            setattr(self, attribute, bin_data(attrVal, attrBin))        

    def Add(self, additive):
        for attribute in self.attrNames:
            attrVal = getattr(self, attribute)
            if len(attrVal) == 0:
                setattr(self, attribute, getattr(additive, attribute))
            else:
                setattr(self, attribute, attrVal + getattr(additive, attribute))
```

Here you have a working example that I have in mind with this structure:

```
S1Vars = {'Number', 'Width', 'Height', 'Energy', 'Charge', 'Time'}
rangeS1 = {'Number': [0, 10, 10], 'Width': [0, 1000, 10], 'Height': [0, 10, 10], 'Energy': [0, 50, 50],\
           'Charge': [0, 20, 20], 'Time': [0, 1300, 1300]} #Bin limits and number of bins

output = "/Users/ansimez/Documents/InvCit/hdf5/histos1_pmaps_001_4670_v0.9.3_20171128_double_kr1300.h5"
h5out = tb.open_file(output, "w")
writer = hist_writer_var(h5out)

files = glob.glob("/Users/ansimez/Documents/InvCit/hdf5/diffusion/4670/pmaps_00*_4670_v0.9.3_20171128_double_kr1300.h5")

S1Bin = MonPar(S1Vars, {attrName: np.arange(attrVal[0], attrVal[1] + (attrVal[1]-attrVal[0]) / attrVal[2], (attrVal[1]-attrVal[0]) / attrVal[2])\
                        for attrName, attrVal in rangeS1.items()})
S1Var = MonPar(S1Vars)
S1Binned = MonPar(S1Vars)

for file in files:
    pmaps = load_pmaps(file)
    S1Var = MonPar(S1Vars)
    for i in pmaps:
        fillS1var(pmaps[i].s1s, S1Var)
    S1Var.BinValues(S1Bin)
    S1Binned.Add(S1Var)

for attribute in S1Binned.attrNames:    
    attrValue = getattr(S1Binned, attribute)
    writer('HIST', 'S1_'+attribute, attrValue[0], shift_to_bin_centers(getattr(S1Bin, attribute)), attrValue[1])
``` 
This is what I currently have in a notebook,  the binning and range are set at the beginning. In S1Bin the bin arrays are stored, in S1Var we store the values, it is created from 0 at the beginning of each file so it does not grow too big and in S1Binned we accumulate the binned values. The fillS1Var just fills the S1 fields with the pmap information:

```
def fillS1var (s1s, S1Var):
    S1Var.Number.append(len(s1s))
    for s1 in s1s:
        S1Var.Width .append(s1.width)
        S1Var.Height.append(s1.height)
        S1Var.Energy.append(s1.total_energy)
        S1Var.Charge.append(s1.total_charge)
        S1Var.Time  .append(s1.time_at_max_energy)
``` 

So this would be the proposal, comments? Are you ok with a scheme like this or should it be done in another way? As I said, I really have no idea of what I'm doing so I'm prepared for the stones if something is a terrible bad idea.

Aretno committed:
Add a general histogram writer

Aretno committed:
Add and test Histogram and HistoManager classes.

Aretno committed:
Add Histogram related functions

Added plot, save histogram an join/sum functions.

Aretno committed:
Add and test monitoring functions for pmaps and rwf

Aretno committed:
Problem in test with - char

Aretno committed:
Cosmetics

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,194 @@
+            data_weights = np.ones (len(additive.T))
+        else:
+            data_weights = np.array(data_weights)
+
```
A few things:
1. The `!=` comparison is to check whether `data_weights` is empty? If so, then just check its size. An invalid length of that sequence should raise an error, I guess.
2. The default value for `data_weights` should be an empty tuple, as the list object is mutable.
3. You can use `a = some_value if condition else other_value` for assignments of this kind. If it is too long, leave it as it is, but put the code in the same line as the `if`/`else`.
4. I find more elegant `additive.shape[1]` than `len(additive.T)`

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,194 @@
+            self.errors = np.asarray(errors)
+        else:
+            self.errors = np.sqrt(self.data)
+
```
Same as before. `a = some_value if condition else other_value` or put the code in the same line as the if/else.

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,194 @@
+    def __add__ (self, other):
+        if other is None:
+            return self
+        if len(self.bins) != len(other.bins) or not np.all(a == b for a, b in zip(self.bins, other.bins)):
```
hide this `if` in a function `_check_valid_binning`

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,194 @@
+            if histoname in self.histos:
+                self[histoname].fill(np.asarray(additive))
+            else:
+                print("Histogram with name {} does not exist".format(histoname))
```
You can use format strings, if you prefer (`f""`)

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,194 @@
+            else:
+                print("Histogram with name {} does not exist".format(histoname))
+
+    def save_to_file(self, file_out, mode='w', group='HIST'):
```
Did we finally agreed on leaving this as a method rather than as a standalone function? I feel now more comfortable with a function that lives in `hist_io`

Aretno replied:
I thought so but I'll see if I move it to hist_io then.

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,194 @@
+        with tb.open_file(file_out, mode, filters=tbl.filters('ZLIB4')) as h5out:
+            writer = hist_writer_var(h5out)
+            for histoname, histo in self.histos.items():
+                writer(group, histoname, histo.data,   histo.bins, histo.out_range,
```
new line after `histoname,`

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +27,36 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+                   out_of_range: 'np.array lenght=2 with events out of range',
+                   errors      : 'np.array with bins uncertainties'          ,
+                   labels      : 'list with labels of the histogram'):
+
```
align `)` with the `,`s

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +27,36 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+        except tb.NoSuchNodeError: hist_group = file.create_group(file.root, group_name)
+
+        if table_name in hist_group:
+            raise ValueError("Histogram {} already exists".format(table_name))
```
you can use format strings, if you prefer

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +27,36 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+        add_carray       (hist_group, table_name + '_outRange'     , out_of_range)
+        add_carray       (hist_group, table_name + '_errors'       , errors      )
+        file.create_array(hist_group, table_name + '_labels'       , labels      )
+
```
remove extra spaces before `,`s

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +27,36 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+        if table_name in hist_group:
+            raise ValueError("Histogram {} already exists".format(table_name))
+
+        vlarray = file.create_vlarray(hist_group, table_name + '_bins', atom=tb.Float64Atom(shape = ()), filters=tbl.filters(compression))
```
split the `vlarray` line so it is shorter?

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +27,36 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+        add_carray       (hist_group, table_name                   , entries     )
+        add_carray       (hist_group, table_name + '_outRange'     , out_of_range)
+        add_carray       (hist_group, table_name + '_errors'       , errors      )
+        file.create_array(hist_group, table_name + '_labels'       , labels      )
```
Why is different the labels array than the other ones?

Aretno replied:
If I recall correctly, because it is not a numpy array. Doing a numpy array of that so the add_carray function works would return a numpy array of dtype object I believe.

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +27,36 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+        array_atom  = tb  .Atom.from_dtype(var.dtype)
+        array_shape = var .shape
+        entry       = file.create_carray(hist_group, table_name, atom=array_atom,
+                                         shape=array_shape, filters=tbl.filters(compression))
```
put each keyword argument in a different line

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+from .. icaro.hst_functions  import shift_to_bin_centers
+
+
+def create_histomanager_from_dicts(histobins_dict, histolabels_dict, init_fill_dict={}):
```
`init_fill_dict` will have to be `None` by default, as the dictionary is mutable.

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    histo_manager = HistoManager()
+    for histotitle, histobins in histobins_dict.items():
+        histo_manager.new_histogram(Histogram(histotitle, histobins, histolabels_dict[histotitle],
```
I would put each argument in a different line, I think

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+def get_histograms_from_file(file_input, group_name='HIST'):
+    histo_manager = HistoManager()
+    with tb.open_file(file_input, "r") as h5in:
+        name_sel = lambda x: (    ('bins'     not in x)
```
use a `def` statement rather than a `lambda`

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+                              and ('outRange' not in x))
+        histogram_list = []
+        group = getattr(h5in.root, group_name)
+        for histoname in filter(name_sel, list(group._v_children)):
```
do you need to wrap `group._v_children` into a list? If it is an iterable, it is not  needed

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+    histofiles = List of strings with the filenames to be summed.
+    join_file  = String. If passed, saves the resulting HistoManager to this path.
+    """
+    if len(histofiles)<1:
```
`if not histofiles`

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+        final_histogram_manager = join_histo_managers     (final_histogram_manager, added_histograms)
+
+    if join_file is not None:
+        final_histogram_manager.save_to_file(join_file)
```
The function `Histomanager.save_to_file` takes extra keyword arguments, you may want to accept them also as parameters of this function

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+    return final_histogram_manager
+
+
+def plot_histograms_from_file(histofile, histonames='all', group_name='HIST', plot_errors=False, out_path=None):
```
Do we want these plotting functions in this module? Probably better somewhere else, but not sure.

gonzaponte replied:
Thoughts on this?

gonzaponte commented on invisible_cities/reco/histogram_functions.py:
```diff
@@ -0,0 +1,218 @@
+    """
+    if b == 0:
+        return -100.
+    return 100. * a / b
```
`return 100 * a / b if b else -100`

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,143 @@
+            aux_histogram.data      =         histo1.data        + histo2.data
+            aux_histogram.errors    = np.sqrt(histo1.errors ** 2 + histo2.errors ** 2)
+            aux_histogram.out_range =         histo1.out_range   + histo2.out_range
+
```
I prefer `true_histogram` rather than  `aux_histogram`

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,143 @@
+    histograms_dict  = {}
+    print(bins)
+    for i, bins_element in enumerate(bins):
+        print(bins_element)
```
Don't forget to remove debug prints! :)

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,143 @@
+    print(bins)
+    for i, bins_element in enumerate(bins):
+        print(bins_element)
+        title =    "Histo_{}" .format(i)
```
format strings would be nice here (more readable)

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,143 @@
+    assert len(joined_histogram_manager1.histos) == len(joined_histogram_manager2.histos)
+    for histoname in joined_histogram_manager1.histos:
+        assert_histogram_equality(joined_histogram_manager1[histoname], joined_histogram_manager2[histoname])
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,143 @@
+
+    file_out_aux = os.path.join(output_tmpdir, 'test_save_histogram_manager_1.h5')
+    histogram_manager.save_to_file(file_out_aux)
+
```
`file_out_test`?

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+            var_bins  [k.replace("_bins", "")]   = [ np.linspace(v[0], v[1], v[2] + 1) ]
+        elif "_labels" in k:
+            var_labels[k.replace("_labels", "")] = v
+
```
- put the `var_*` in the same line as the `if`s.
- align the `, "")`
- remove the spaces surrounding the `np.linspace`

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+    var_labels    ['S2_Time_S2_Energy']   = var_labels['S2_Time']   + var_labels['S2_Energy']
+    var_bins      ['S2_Energy_S1_Energy'] = var_bins  ['S2_Energy'] + var_bins  ['S1_Energy']
+    var_labels    ['S2_Energy_S1_Energy'] = var_labels['S2_Energy'] + var_labels['S1_Energy']
+    var_bins      ['S2_XYSiPM']           = var_bins  ['S2_XSiPM']  + var_bins['S2_XSiPM']
```
alignment

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+    var_bins      ['S2_XYSiPM']           = var_bins  ['S2_XSiPM']  + var_bins['S2_XSiPM']
+    var_labels    ['S2_XYSiPM']           = var_labels['S2_XSiPM']  + ['Y (mm)']
+
+    del var_bins['S2_XSiPM']
```
why is there just the `x` variable, but not the `y` one?

Aretno replied:
In the config_dict I do not pass any YSiPM bining, just get the same as XSiPM therefore in the end I only need to erase the X one. The XY plot I'm generating does not need a different binning in Y (nor the possibility, I would say).

gonzaponte replied:
Ok. Can we name it `whatever_XY`?

Aretno replied:
In the end I just added a Y field and erased both of them, thought it was clearer.

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+    del var['S2_SingleS1_Energy']
+
+    return var
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+            var = fill_pmap_var(pmaps[pi], SiPM_db)
+            histo_manager.fill_histograms(var)
+    return histo_manager
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+    var_bins, var_labels = pmap_bins(config_dict)
+    histo_manager = histf.create_histomanager_from_dicts(var_bins, var_labels)
+    SiPM_db = dbf.DataSiPM(run_number)
+
```
align equals

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+            histo_manager.fill_histograms(var)
+    return histo_manager
+
+def histogram_pmap_variable_creator(run_number):
```
These functions will be introduced for the liquid city. They are not needed here.

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+        elif "_labels" in k:
+            var_labels[k.replace("_labels", "")] = v
+
+    return var_bins, var_labels, config_dict['n_baseline']
```
same comments as for `pmap_bins`

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+    rms = np.std (rwf, axis=1)[:, np.newaxis]
+    var_dict[sensor_type + '_Baseline']   .extend(bls.flatten())
+    var_dict[sensor_type + '_BaselineRMS'].extend(rms.flatten())
+
```
`array[:, np.newaxis].flatten()` is the same as `array` :).

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+    var_dict[sensor_type + '_Baseline']   .extend(bls.flatten())
+    var_dict[sensor_type + '_BaselineRMS'].extend(rms.flatten())
+
+def fill_rwf_histos(in_path, config_dict):
```
two blank lines

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,222 @@
+            for evt in range(nevt):
+                fill_rwf_var(pmtrwf [evt,:,0:n_baseline], var,  "PMT")
+                fill_rwf_var(sipmrwf[evt]               , var, "SiPM")
+
```
`pmtrwf[evt, :, :n_baseline]`

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+        assert np.allclose(var_dict['S2_XSiPM'] [counter:counter + nsipm], data_sipm.X.values[speak.sipms.ids].tolist())
+        assert np.allclose(var_dict['S2_YSiPM'] [counter:counter + nsipm], data_sipm.Y.values[speak.sipms.ids].tolist())
+        counter += nsipm
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+        assert var_dict['S1_Energy_S1_Charge'][1,i] == speak.total_charge
+        assert var_dict['S1_Time_S1_Energy']  [0,i] == speak.time_at_max_energy /units.mus
+        assert var_dict['S1_Time_S1_Energy']  [1,i] == speak.total_energy
+
```
`[a, b]` rather than `[a,b]`

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+        assert np.allclose(var_dict['S2_XYSiPM'][0, counter:counter + len(sipm_ids)], data_sipm.X.values[speak.sipms.ids].tolist())
+        assert np.allclose(var_dict['S2_XYSiPM'][1, counter:counter + len(sipm_ids)], data_sipm.Y.values[speak.sipms.ids].tolist())
+        counter += len(sipm_ids)
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+                 'S2_Energy': [    0., 7.5e3,          15.0e3 ],
+                 'S2_XSiPM' : [ -100.,  -50.,  0.,  50., 100. ],
+                 'S2_Time'  : [    5.,  6.25, 7.5, 8.75,  10. ]}
+
```
- Don't use a dot in the numbers when not necessary. In python 3, `1` behaves as `1.0` when required.
- Align decimal points (those that are actually necessary).
- I think is better to put
```
[x, y, z      ]
[a, b, c, d, e]
```
rather than
```
[x, y,     , z]
[a, b, c, d, e]
```
unless there is a symmetry I am not seeing.

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+    out_bins, out_labels = monf.pmap_bins(test_dict)
+
+    list_var = [ 'S1_Energy', 'S1_Width', 'S1_Time', 'S2_Energy', 'S2_Time' ]
+
```
remove spaces after/before `[` and `]`

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+        assert k in [ 'S1_Energy', 'S1_Width', 'S1_Time', 'S2_Energy', 'S2_Time',
+                      'S1_Energy_S1_Width', 'S1_Time_S1_Energy', 'S2_Time_S2_Energy',
+                      'S2_Energy_S1_Energy', 'S2_XYSiPM' ]
+
```
double blank line

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+    for k in out_bins:
+        assert k in [ 'S1_Energy', 'S1_Width', 'S1_Time', 'S2_Energy', 'S2_Time',
+                      'S1_Energy_S1_Width', 'S1_Time_S1_Energy', 'S2_Time_S2_Energy',
+                      'S2_Energy_S1_Energy', 'S2_XYSiPM' ]
```
remove spaces after/before `[` and `]`

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+                        'S2_IdSiPM_labels' : [              'SiPM id' ],
+                        'S2_QSiPM_labels'  : [    'SiPM charge (pes)' ],
+                        'S2_XSiPM_labels'  : [               'X (mm)' ]}
+
```
- remove spaces after/before `[` and `]`
- Are the raw strings necessary?
- Check the spaces after the first comma in the bins section.
- You can use Unicode µ (alt + m) if it is handy and does not create problems.

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+        assert np.allclose(v.out_range, test_histo.histos[k].out_range)
+        assert np.allclose(v.errors   , test_histo.histos[k].errors   )
+        assert             v.title   == test_histo.histos[k].title
+        for i, label in enumerate(v.labels):
```
If they are both lists of strings you can simply say `assert v.labels == test_histo.histos[k].labels` rather than using an explicit loop. Same for bins

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+def test_fill_rwf_var():
+    var_dict = defaultdict(list)
+    pmt_waveforms  = np.random.uniform(0., 10., size=(  12, 10000))
+    sipm_waveforms = np.random.uniform(0., 10., size=(1792, 10000))
```
remove unnecessary `.`s

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+    assert np.allclose(var_dict['PMT_BaselineRMS'] , np.std (pmt_waveforms , axis=1)[:, np.newaxis].flatten())
+    assert np.allclose(var_dict['SiPM_Baseline']   , np.mean(sipm_waveforms, axis=1)[:, np.newaxis].flatten())
+    assert np.allclose(var_dict['SiPM_BaselineRMS'], np.std (sipm_waveforms, axis=1)[:, np.newaxis].flatten())
+
```
`[:, np.newaxis].flatten()` equals doing nothing

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+                 'PMT_BaselineRMS_bins'    : [    0.,   10., 100 ],
+                 'SiPM_Baseline_bins'      : [    0.,  100., 100 ],
+                 'SiPM_BaselineRMS_bins'   : [    0.,   10., 100 ],
+
```
remove spaces before/after `[`/`]` and check the spaces before `:`

Aretno replied:
The spaces before `:` are due to being aligned with `'SiPM_BaselineRMS_labels'`

gonzaponte replied:
ok, I didn't realize

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+    assert np.allclose(out_bins['SiPM_Baseline']   , [ np.linspace(aux[0], aux[1], aux[2] + 1)])
+    aux = test_dict['SiPM_BaselineRMS_bins']
+    assert np.allclose(out_bins['SiPM_BaselineRMS'], [ np.linspace(aux[0], aux[1], aux[2] + 1)])
+
```
- rename all `aux` to a better name (`bins`? actually maybe even better `min, max, nbin` instead of a single variable)
- spaces before/after `[`/`]`

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+                        'PMT_BaselineRMS_bins'    : [    0.,   10., 100 ],
+                        'SiPM_Baseline_bins'      : [    0.,  100., 100 ],
+                        'SiPM_BaselineRMS_bins'   : [    0.,   10., 100 ],
+
```
spaces and decimal points

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -0,0 +1,296 @@
+        for i, label in enumerate(v.labels):
+            assert           label   == test_histo.histos[k].labels[i]
+        for i, bins in enumerate(v.bins):
+            assert np.allclose(bins,    test_histo.histos[k].bins[i])
```
if they are lists use `assert list_a == list_b`, same as before.

gonzaponte reviewed (commented):
It took me a while but finally got there. Mostly ok, but some small cosmetics details still to be fixed. Also, please check that there are not unused imports in the files you've introduced.

Aretno committed:
Modifications according to the PR review

Aretno committed:
Move get_histograms_from_file to hist_io

Aretno committed:
Added the number of sensors in wvf plots

Aretno committed:
Temporal script to run monitoring

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,173 @@
+import numpy  as np
+import tables as tb
+
+from .. reco       import tbl_functions as tbl
```
extra spaces

gonzaponte commented on invisible_cities/evm/histos.py:
```diff
@@ -0,0 +1,173 @@
+        data_weights = Array or list with weights of the data.
+        """
+        additive = np.array(additive)
+        data_weights = np.ones(len(additive.T)) if data_weights is None else np.array(data_weights)
```
align equals

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -1,6 +1,14 @@
+from .. reco    import tbl_functions as tbl
+
+from .. evm.histos import HistoManager
+from .. evm.histos import Histogram
```
align imports

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +35,91 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+
+        vlarray = file.create_vlarray(hist_group, table_name + '_bins',
+                                      atom=tb.Float64Atom(shape=()),
+                                      filters=tbl.filters(compression))
```
align equals with spaces surrounding them

gonzaponte commented on invisible_cities/io/hist_io.py:
```diff
@@ -27,3 +35,91 @@ def write_hist(histo : 'np.array: RWF, CWF, SiPM'):
+        entry       = file.create_carray(hist_group, table_name,
+                                         atom=array_atom,
+                                         shape=array_shape,
+                                         filters=tbl.filters(compression))
```
align equals with spaces surrounding them

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,128 @@
+from .. evm.histos_test     import assert_histogram_equality
+from .. evm.histos_test     import histograms_lists
+from .. evm.histos_test     import bins_arrays
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,128 @@
+
+    assert len(list_of_histograms) == len(joined_histogram_manager.histos)
+    for histoname, histogram in joined_histogram_manager.histos.items():
+        histo1 = histogram_manager[histoname]
```
align equals

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,128 @@
+
+    for histoname, histogram in joined_histogram_manager.histos.items():
+        assert histoname in list_of_names
+        if (histoname in histogram_manager1.histos) and (histoname in histogram_manager2.histos):
```
maybe in two lines like:
```
if (histoname in histogram_manager1.histos and
    histoname in histogram_manager2.histos    ):
```

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,128 @@
+    histograms_dict  = {}
+    for i, bins_element in enumerate(bins):
+        title =    f"Histo_{i}"
+        labels = [ f"Xlabel_{i}", f"Ylabel_{i}" ]
```
align equals and spaces before/after `[`/`]`

gonzaponte commented on invisible_cities/reco/histogram_functions_test.py:
```diff
@@ -0,0 +1,128 @@
+
+    assert len(joined_histogram_manager1.histos) == len(joined_histogram_manager2.histos)
+    for histoname in joined_histogram_manager1.histos:
+        assert_histogram_equality(joined_histogram_manager1[histoname], joined_histogram_manager2[histoname])
```
split into two lines and align

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,203 @@
+        var_bins  ['S1_Energy_' + param] = var_bins  ['S1_Energy'] + var_bins  [param]
+        var_labels['S1_Energy_' + param] = var_labels['S1_Energy'] + var_labels[param]
+    var_bins      ['S1_Time_S1_Energy']  = var_bins  ['S1_Time'] + var_bins  ['S1_Energy']
+    var_labels    ['S1_Time_S1_Energy']  = var_labels['S1_Time'] + var_labels['S1_Energy']
```
align plus signs

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,203 @@
+    return var_bins, var_labels
+
+
+def fill_pmap_var_1d(speaks, var_dict, ptype, DataSiPM=None):
```
I didn't realize before, but the peak has its own type, so `ptype` is redundant. Instead of `if ptype == "S2"` you can say `if isinstance(speak, S2)` where `S2` comes from `evm.pmaps`.

gonzaponte replied:
You can get the string representation with `type(speak).__name__`

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,203 @@
+    param_list = ['Width', 'Height', 'Charge']
+    for param in param_list:
+        var_dict[ptype + '_Energy_' + ptype + '_' + param] = np.array([var_dict[ptype + '_Energy'], var_dict[ptype + '_' + param]])
+    var_dict    [ptype + '_Time_' + ptype + '_Energy']     = np.array([var_dict[ptype + '_Time']  , var_dict[ptype + '_Energy']])
```
align plus signs

gonzaponte commented on invisible_cities/reco/monitor_functions.py:
```diff
@@ -0,0 +1,203 @@
+        elif "_labels" in k: var_labels[k.replace("_labels", "")] = v
+
+    return var_bins, var_labels, config_dict['n_baseline']
+
```
two blank lines

gonzaponte commented on invisible_cities/reco/olivia.py:
```diff
@@ -0,0 +1,62 @@
+
+data_types = ['rwf', 'pmaps'  ]
+sources    = ['kr' , 'generic']
+config_files = {'pmaps_kr'      : os.path.join(ICDIR, 'config/PmapsKrConfig.json')     ,
```
The city shouldn't know anything about the configuration files available. It should be itself an argument. Actually, the config file should be the only parameter and follow somehow the structure we already have.

gonzaponte commented on invisible_cities/reco/olivia.py:
```diff
@@ -0,0 +1,62 @@
+    out_path   = sys.argv[2]
+    run_number = int(sys.argv[3])
+    data_type  = sys.argv[4]
+    source     = sys.argv[5]
```
align `sys.`

gonzaponte commented on invisible_cities/reco/olivia.py:
```diff
@@ -0,0 +1,62 @@
+    source     = sys.argv[5]
+
+    if data_type not in data_types:
+        print('Error: Data type {} is not recognized.\n'.format(data_type))
```
format strings
remove `\n`

gonzaponte commented on invisible_cities/reco/olivia.py:
```diff
@@ -0,0 +1,62 @@
+
+    if source not in sources:
+        print('Error: Source {} is not recognized.\n'.format(source))
+        print('Supported sources: {}'.format(sources))
```
format strings
remove `\n`

gonzaponte commented on invisible_cities/reco/olivia.py:
```diff
@@ -0,0 +1,62 @@
+        save_histomanager_to_file(histo_manager, out_path)
+
+    elif data_type == 'pmaps':
+        with open(config_files[data_type+'_'+source]) as config_file:
```
spaces surrounding `+`

gonzaponte reviewed (commented):
Minor cosmetics + reorganize olivia arguments

Aretno committed:
Adapt code to reviewer petitions.

Aretno committed:
Modified style of 1D histograms (to match paper style)

Aretno committed:
Variable name change to common criteria.

Aretno committed:
Moved plotting functions to icaro folder

Aretno committed:
Tests for hist_io

Aretno committed:
Added individual PMT plots

Aretno committed:
Added fill_pmt_var test

gonzaponte commented on invisible_cities/io/hist_io_test.py:
```diff
@@ -0,0 +1,105 @@
+            saved_labels = [str(label)[2:-1].replace('\\\\', '\\') for label in getattr(file_group, histoname + "_labels")[:]]
+
+            assert histoname in file_group
+            assert np.all     (a == b for a, b in zip(histogram.bins, getattr(file_group, histoname + "_bins")[:]))
```
You need to ensure the length of  `histogram.bins` and `getattr(file_group, histoname + "_bins")[:]`matches. `zip` truncates the output to the shorter of the inputs.

gonzaponte replied:
I guess we have discussed it a few times, but why is it not enough `np.allclose(histogram.bins, getattr(file_group, histoname + "_bins")[:]`?

Aretno replied:
For 2D plots the bins don't have de same dimension so the bins are a array of dtype=object which I think fails when using the allclose.

gonzaponte commented on invisible_cities/io/hist_io_test.py:
```diff
@@ -0,0 +1,105 @@
+            assert np.allclose(histogram.data     , getattr(file_group, histoname              )[:])
+            assert np.allclose(histogram.errors   , getattr(file_group, histoname + "_errors"  )[:])
+            assert np.allclose(histogram.out_range, getattr(file_group, histoname + "_outRange")[:])
+            assert             histogram.labels  == saved_labels
```
Align `histogram`, `getattr` and `)[:])`

gonzaponte commented on invisible_cities/io/hist_io_test.py:
```diff
@@ -0,0 +1,105 @@
+            assert np.allclose(histogram.errors   , getattr(file_group, histoname + "_errors"  )[:])
+            assert np.allclose(histogram.out_range, getattr(file_group, histoname + "_outRange")[:])
+            assert             histogram.labels  == saved_labels
+
```
same as above

gonzaponte commented on invisible_cities/io/hist_io_test.py:
```diff
@@ -0,0 +1,105 @@
+
+    file_out   = os.path.join      (output_tmpdir, 'test_save_histogram_manager.h5')
+    write_mode = write_mode.replace('w','')
+    write_mode = write_mode.replace('a','')
```
It is better to exclude `"w"` and `"a"` from the input. Try `text(characters, min_size=1, max_size=1).filter(lambda x: x not in ["w", "a"])`

gonzaponte replied:
I meant, add the `filter` to the input and remove the `replace` calls here :)

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -38,7 +38,6 @@ def test_fill_pmap_var_1d(pmaps):
         assert     var_dict['S2_Energy']  [i] == speak.total_energy
         assert     var_dict['S2_Width']   [i] == speak.width / units.mus
         assert     var_dict['S2_Height']  [i] == speak.height
-        assert     var_dict['S2_Energy']  [i] == speak.total_energy
```
why do you remove that?

Aretno replied:
It was duplicated (line 38).

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -93,6 +92,28 @@ def test_fill_pmap_var_2d(pmaps):
+        for j in range(npmts):
+            assert var_dict['PMT' + str(j) + '_S2_Energy'][i] == energies[j]
+            assert var_dict['PMT' + str(j) + '_S2_Height'][i] == heights [j]
+            assert var_dict['PMT' + str(j) + '_S2_Time'  ][i] == times   [j]
```
It is shorter to use format strings

gonzaponte reviewed (commented):
Much shorter review than in previous occasions! 😄

Aretno committed:
Review comments

Aretno committed:
Plot 3d histogramas as 2d profiles

Aretno committed:
Added monitoring plots for kdst

Aretno committed:
Added options in order to plot reference histograms

Added extra arguments for reference histograms (only 1d), histograms
are now normalized (optional) for the sake of comparison, stats
ploting is now an option.

Aretno committed:
Added kdst related config and test histofile

gonzaponte commented on invisible_cities/reco/monitor_functions_test.py:
```diff
@@ -152,38 +154,11 @@ def test_pmap_bins():
+    assert_bins_and_labels_ndim('S1_Time_S1_Energy'  , ['S1_Time'  , 'S1_Energy'], out_bins, out_labels, test_bins, test_dict)
+    assert_bins_and_labels_ndim('S2_Time_S2_Energy'  , ['S2_Time'  , 'S2_Energy'], out_bins, out_labels, test_bins, test_dict)
+    assert_bins_and_labels_ndim('S2_Energy_S1_Energy', ['S2_Energy', 'S1_Energy'], out_bins, out_labels, test_bins, test_dict)
+    assert_bins_and_labels_ndim('S2_XYSiPM'          , ['S2_XSiPM' , 'S2_YSiPM' ], out_bins, out_labels, test_bins, test_dict)
```
Like!

gonzaponte reviewed (approved):
The time has come. This PR is approved.

Lots of ugly (although necessary) functions to implement data monitoring have been added.
A review and refactor of them will probably be necessary at some point, but it is good to go.
Hard work from @Aretno!!!

Aretno committed:
Correct RMS
